src/query-executor/query-executor-base.ts

```diff
@@ -68,6 +68,19 @@ export abstract class QueryExecutorBase implements QueryExecutor {
     })
   }
 
+  async *stream<R>(
+    compiledQuery: CompiledQuery,
+    queryId: QueryId
+  ): AsyncIterableIterator<QueryResult<R>> {
+    const connection = await this.provideConnection(
+      async (connection) => connection
```

Comment:
This code releases the connection to the pool before returning the connection. So you are actually using a released connection here.

You need to do the work inside the `provideConnection`'s callback argument. Only once the stream is done, should you release the connection back to the pool.

Reply:
I can't really do the awaiting inside of `provideConnection` callback in this case because of `yield` (can't return async generator as `T` from `provideConnection` because connection has to be active during lifetime of async generator). I have 2 ideas though:
1. I can prevent `provideConnection` from returning until the connection is not needed anymore (iterator is done).
2. Add "unmanaged" `getConnection` to `ConnectionProvider` interface:
```ts
export interface ConnectionProvider {
  provideConnection<T>(
    consumer: (connection: DatabaseConnection) => Promise<T>
  ): Promise<T>

  getConnection(): Promise<{ connection: DatabaseConnection, release: () => void }>
}
```

Reply:
Whichever is fine. I can refactor this if I find a better solution. (I probably won't,☺️)